053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,457 @@
+    for {
+        var evt event.Envelope
+        if err := dec.Decode(&evt); err == io.EOF {
+            break
+        } else if err != nil {
+            return nil, err
+        }
+        events = append(events, evt)
+    }
+    return events, nil
+}
+```
+
+The point I'm trying to illustrate is that the `EventStore` hooks allow you to
+do some pretty cool things when you compose `EventStores`. If you can think of
+any hooks that are missing that could support other useful functionality, I'm
+all ears!
+
+## <a name="notifications">`Notifications`</a>
+
+As I mentioned before, while you can iterate over a completed build's events by
+looping over `EventStore.Get`, this doesn't work so well for builds that are
+still producing events, since (in general) there's no way to know when new
+events come in (besides polling). The solution that I came up with is to use
+the Postgres notification bus to notify `db.Build.Events()` of new build
+event(s). It feels a bit weird to rely on Postgres for this even when using an
+external store that shouldn't need to touch Postgres - I left an open question
+about whether people feel this is a bad idea, and another question about
+whether there are use cases where `EventStores` would benefit from using their
+own native pub/sub system (e.g. MongoDB has a `collection.watch()`
+functionality).
+
+Note that `EventStore` implementations don't need to think about notifications.
+This all happens at the DB layer:
+
+```go
+func (b *build) SaveEvent(event atc.Event) error {
+    // use the `EventStore` to save
+    err := b.eventStore.Put(b, event)
+    if err != nil {
+        return err
+    }
+    // ...but notify the main Postgres DB's notification bus.
+    return b.conn.Bus().Notify(buildEventsNotificationChannel(b.ID()))
+}
+```
+
+```go
+func (b *build) Events(from uint) (EventSource, error) {
+    notifier, err := newConditionNotifier(b.conn.Bus(), buildEventsNotificationChannel(b.ID()), func() (bool, error) {
+        return true, nil
+    })
+    if err != nil {
+        return nil, err
+    }
+
+    return newBuildEventSource(
+        b,
+        b.conn,
+        b.eventStore, // fetch events from the EventStore using EventStore.Get(...)
+        notifier,     // but subscribe to the Postgres notification bus to know when new events arrive
+        from,
+    ), nil
+}
+```
+
+I've toyed with the idea of having an optional interface that `EventStore`
+implementations can implement that would overwrite this default...something
+like:
+
+```go
+type EventNotificationBus interface {
+    Notify(build Build) error
+    Listen(build Build) (<-chan struct{}, error)
+}
+```
+
+But I'm not sure if there's much value in that.
+
+
+## <a name="migrations">Migrations</a>
+
+Currently, we have migrations to set up the `build_events` table, as well as
+`TRIGGER`s for creating/deleting `pipeline_build_events_x` and
+`team_build_events_x` tables when pipelines and teams are created/deleted.
```

Comment:
Aside: I wish this were achieved with partitions rather than templated tables.

Reply:
@pivotal-jwinters @vito (I guess?) is this something considered back then?